### PR TITLE
util: optimize the memory usage of the read path for listInDisk

### PR DESCRIPTION
**Pull request by wshwsh12**
### What problem does this PR solve?

Issue Number: ref https://github.com/pingcap/tidb/issues/33877

Problem Summary:

Optimize the memory usage of the read path for list in disk.

### Check List

Tests 

- [ ] Unit test
- [ ] Integration test
- [x] Manual test (add detailed scripts or steps below)
Workload: A big table with 5kw lines 1, a small table with 1 lines 1.
SQL: desc analyze   select * from small_table where exists (select a from v5kw big_table where small_table.b=big_table.b);
Plan: With memory quota 1GB
```
root@127.0.0.1:test_xhy> desc    select * from small_table where exists (select a from v5kw big_table where small_table.b=big_table.b);
+----------------------------------+-------------+-----------+-------------------+----------------------------------------------------------------------+
| id                               | estRows     | task      | access object     | operator info                                                        |
+----------------------------------+-------------+-----------+-------------------+----------------------------------------------------------------------+
| HashJoin_15                      | 0.80        | root      |                   | semi join, equal:[eq(test_xhy.small_table.b, test_xhy.base_table.b)] |
| ├─Selection_19(Build)            | 40000000.00 | root      |                   | not(isnull(test_xhy.base_table.b))                                   |
| │ └─Limit_20                     | 50000000.00 | root      |                   | offset:0, count:50000000                                             |
| │   └─TableReader_24             | 50000000.00 | root      |                   | data:Limit_23                                                        |
| │     └─Limit_23                 | 50000000.00 | cop[tikv] |                   | offset:0, count:50000000                                             |
| │       └─TableFullScan_22       | 50000000.00 | cop[tikv] | table:base_table  | keep order:false                                                     |
| └─TableReader_18(Probe)          | 1.00        | root      |                   | data:Selection_17                                                    |
|   └─Selection_17                 | 1.00        | cop[tikv] |                   | not(isnull(test_xhy.small_table.b))                                  |
|     └─TableFullScan_16           | 1.00        | cop[tikv] | table:small_table | keep order:false, stats:pseudo                                       |
+----------------------------------+-------------+-----------+-------------------+----------------------------------------------------------------------+
9 rows in set (0.00 sec)
```
Before this pr: TiDB use **60GB+** and OOM.
After this pr: We can see the GC line is 5.6GB, so the memory usage after each GC is 2.8GB
image

- [ ] No code

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

Please refer to [Release Notes Language Style Guide](https://pingcap.github.io/tidb-dev-guide/contribute-to-tidb/release-notes-style-guide.html) to write a quality release note.

```release-note
None
```

### CODE REVIEW

File: util/chunk/disk.go [hunk 1]
```diff
@@ -39,6 +39,8 @@ type ListInDisk struct {
 
 	dataFile   diskFileReaderWriter
 	offsetFile diskFileReaderWriter
+
+	chk *Chunk
```
Comment (XuHuaiyu) [inline util/chunk/disk.go:43]: Add a comment for this attribute.

File: util/chunk/disk.go [hunk 2]
```diff
@@ -378,35 +387,28 @@ func convertFromRow(row Row, reuse *diskFormatRow) (format *diskFormatRow) {
 	return
 }
 
-// toMutRow deserializes diskFormatRow to MutRow.
-func (format *diskFormatRow) toMutRow(fields []*types.FieldType) MutRow {
-	chk := &Chunk{columns: make([]*Column, 0, len(format.sizesOfColumns))}
+// toRow deserializes diskFormatRow to Row.
+func (format *diskFormatRow) toRow(fields []*types.FieldType, chk *Chunk) (Row, *Chunk) {
+	if chk == nil || chk.IsFull() {
```
Comment (guo-shaoge) [inline util/chunk/disk.go:392]: Just a quesction, `chk == nil` maybe unnecessary? Because sync.Pool will always new a chunk.

Comment (wshwsh12) [reply util/chunk/disk.go:392]: Check `chk==nil` is more safe.. For example, the code in test cases doesn't have the sync.Pool and doesn't need to reuse.

**Review by guo-shaoge — APPROVED** [review]

**Review by XuHuaiyu — APPROVED** [review]

**Pull request merged**